packages/caravan-health/src/privacy.ts

```diff
@@ -0,0 +1,204 @@
+The score is further influenced by several factors such as address reuse, 
+address types and UTXO set fingerprints etc.
+
+More on the algorithms for scoring privacy: <TODO : Add Link to Blog Post>
+*/
+
+// A normalizing quantity that increases the score by a certain factor in cases of self-payment.
+// More about deniability : https://www.truthcoin.info/blog/deniability/
+const DENIABILITY_FACTOR = 1.5;
+
+/*
+The p_score is calculated by evaluating the likelihood of self-payments, the involvement of 
+change outputs and the type of transaction based on number of inputs and outputs.
+
+We have 5 categories of transaction type
+- Sweep Spend
+- Simple Spend
+- UTXO Fragmentation
+- Consolidation
+- CoinJoin
+*/
+export function privacyScoreOnIO(transaction: any, client: BlockchainClient): number {
+  const numberOfInputs: number = transaction.vin.length;
+  const numberOfOutputs: number = transaction.vout.length;
+
+  let score: number;
+
+  if (numberOfInputs === 1) {
+    if (numberOfOutputs === 1) {
+      // Sweep Spend (No change Output)
+      // #Input = 1, #Output = 1
+      score = 1 / 2;
+    } else if (numberOfOutputs === 2) {
+      // Simple Spend (Single change output)
+      // #Input = 1, #Output = 2
+      score = 4 / 9;
+    } else {
+      // UTXO Fragmentation
+      // #Input = 1, #Output > 2
+      score = 2 / 3 - 1 / numberOfOutputs;
+    }
+    if (isSelfPayment(transaction,client)) {
+      return score * DENIABILITY_FACTOR;
+    }
+  } else {
+    if (numberOfOutputs === 1) {
+      // Consolidation
+      // #Input >= 2, #Output = 1
+      score = 1 / numberOfInputs;
+
+      // No D.F for consolidation
+    } else {
+      // Mixing or CoinJoin
+      // #Input >= 2, #Output >= 2
+      let x = Math.pow(numberOfOutputs, 2) / numberOfInputs;
+      score = (0.75 * x) / (1 + x);
+      if (isSelfPayment(transaction,client)) {
+        return score * DENIABILITY_FACTOR;
+      }
+    }
+  }
+  return score;
+}
+
+/*
+Determine whether a Bitcoin transaction is a self-payment, meaning the sender and recipient
+are the same entity. To check this, we make an RPC call to the watcher wallet asking for the
+amount that a given address holds. If the call returns a number then it is part of wallet
+otherwise it is not a self payment.
+*/
+function isSelfPayment(transaction: any, client: BlockchainClient): boolean {
+  transaction.vout.forEach(async op => {
+     if(await client.getAddressStatus(op.scriptPubKey.address) === undefined){
+       return false;
+     }
+  })
+  return true;
+}
+
+/* TODO : replace any type to custom types for Transactions and UTXOs*/
+/* 
+In order to score for address reuse we can check the amount being hold by reused UTXOs 
+with respect to the total amount
+*/
+export function addressReuseFactor(utxos: Array<any>): number {
+  let reused_amount : number = 0;
+  let total_amount : number = 0;
+  utxos.forEach((utxo) => {
+    if (utxo.reused) {
```

Comment:
where would this property come from?

Reply:
On running `listunspent` it shows up into the data. I suppose should be coming from the UTXO Type ?
https://developer.bitcoin.org/reference/rpc/listunspent.html

Reply:
Umm Hmm. I see, On the mempool response we don't get this `reused` field. Will need to think of some workaround for this.

Reply:
Right, and according to the docs even with bitcoind it might not be available: "only present if avoid_reuse is set". So the way to do this would be to aggregate a count of transactions with an output to a given address. If it's more than 1 then reused is true.

Reply:
Makes sense. Thanks for clarification.